docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,258 @@
+processes.) In the current implementation, and even previous versions,
+modifying node initialization or injecting custom components required
+copying most of the `node` package, which amounts to requiring users
+to maintain a vendored copy of tendermint for these kinds of use
+cases.
+
+While this is (likely) not tenable in the long term, as users do want
+more modularity, and the current service implementation is brittle and
+difficult to maintain, in the short term it may be possible to delay
+implementation somewhat. Eventually, however, we will need to make the
+`node` package easier to maintain and reason about.
+
+### Generic Service Plugability
+
+We can imagine a system design that exports interfaces (in the Golang
+sense) for all components of the system, to permit runtime dependency
+injection of all components in the system so that users can compose
+tendermint nodes of arbitrary user supplied components.
+
+This is an interesting proposal, and customization would provide
+benefits, this is a huge undertaking particularly with regards to API
+design work, that we do not have scope for at the moment. While this
+is not a goal at the moment, eventually providing support for some
+kinds of plugability may be useful, so the current solution does not
+explicitly forclose the possibility of this alternative.
+
+### Abstract Dependency Based Startup and Shutdown
+
+The proposal on in this document simplifies and makes tendermint node
+initialization more abstract, but the system lacks a number of
+features which daemon/service initialization might provide, such as a
+dependency based system that allows the authors of services to control
+the initialization and shutdown order of components using an
+dependencies to control the ordering.
+
+Such a system would work, by allowing services to declare
+initialization order dependencies to other reactors (by ID, perhaps)
+so that the node could decide the initialization based on the
+dependencys declared by components rather than requiring the node to
+encode this logic directly. 
+
+However, this may be too abstract, complicated, and overwrought, and
+in any case authors of components in the current implementation of
+tendermint *do* need to know about other components, so a dependency
+based system would be unhelpfully abstract at this stage.
+
+## Decisions
+
+- Provide a more flexible internal framework for initializing tendermint
+  nodes to make the initatilization process less hard-coded by the
+  implementation of the node objects.
+
+  - Reactors should not need to expose their interfaces *within* the
+	implementation of the node type, except in the context of some groups of
+	services.
+
+  - This refactoring should be entirely opaque to users.
+
+  - These node initialization changes should not require a
+	reevaluation of the `service.Service` or a generic initialization
+	orchestration framework.
+
+- If required for the 0.35 release, add an "unsafe" way to construct a
```

Comment:
Is it maybe worth noting that this "unsafe" mechanism is intended to be a temporary workaround?

Reply:
Agree. Any custom mempool support we add for the 0.35 series will most likely be temporary and potentially not compliant with future custom mempool support in 0.36 and onwards